payment_gateway_adyen/services/payment_service.py

```diff
@@ -0,0 +1,174 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Akretion (http://www.akretion.com).
+# @author Raphael Valyi <raphael.valyi@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.exceptions import Warning as UserError
+from odoo.tools.translate import _
+from odoo.tools.float_utils import float_round
+from odoo.addons.component.core import Component
+import json
+import logging
+_logger = logging.getLogger(__name__)
+
+try:
+    import Adyen
+except ImportError:
+    _logger.debug('Cannot import Adyen')
+
+
+MAP_SOURCE_STATE = {  # TODO check if complete
+    '[cancel-received]': 'cancel',
+    'Authorised': 'to_capture',
+    'consumed': 'succeeded',
+    'Refused': 'failed',
+    'RedirectShopper': 'pending',
+    '[capture-received]': 'succeeded'}
+
+# decimal points of currencies https://docs.adyen.com/developers/currency-codes
+ZERO_DECIMAL_CURRENCIES = [
+    u'CVE', u'DJF', u'GNF', u'IDR', u'JPY', u'KMF', u'KRW', u'PYG',
+    u'RWF', u'UGX', u'VND', u'VUV', u'XAF', u'XOF', u'XPF',
+],
+
+THREE_DECIMAL_CURRENCIES = [
+    u'BHD', u'JOD', u'KWD', u'LYD', u'OMR', u'TND'
+]
+
+
+class PaymentService(Component):
+    _inherit = 'payment.service'
+    _name = 'payment.service.adyen'
+    _usage = 'gateway.provider'
+    _allowed_capture_method = ['immediately']
+
+    def process_return(self, **params):
+        payload = {}
+        payload["browserInfo"] = {
+            "userAgent": "Mozilla/5.0 (X11; U; Linux i686; en-US; rv:\
+            1.9) Gecko/2008052912 Firefox/3.0",
+            "acceptHeader": "text/html,application/xhtml+xml,\
+            application/xml;q=0.9,*/*;q=0.8"
+        }
+        payload["md"] = params['MD']
+        payload["paResponse"] = params['PaRes']
+        transaction = None
+        try:
+            result = self._get_adyen_client().authorise3d(request=payload)
+            vals = {
+                'state': MAP_SOURCE_STATE[result.message['resultCode']],
+                'data': json.dumps(result.message),
+                }
+
+            transaction = self.env['gateway.transaction'].search([
+                ('external_id', '=', result.message['pspReference']),
+                ('payment_mode_id.provider', '=', 'adyen'),
+                ])
+        except ValueError as e:
+            # NOTE catch orther errors?
+            # do we want to cancel the transaction? Can we do it?
```

Comment:
TODO we need to catch error code correctly and map then to message. Do not forget the error raise here will be visible on the website.
Having such message on front is not possible.

> 
    Received validation error with errorCode: 103, message: CVC is not the right
    length, HTTP Code: 422. Please verify the values provided. Please reach out
    to support@adyen.com if the problem persists, providing the PSP reference:
    8515278684882800'

Reply:
so as we have seen the HEAD code of the Adyen module fixed the exception management. Let's propose a patch storing the errorCode and we can indeed re-catch the errors properly. Coming in next commits...